### PR TITLE
Reference guide view page

### PR DESCRIPTION
**Pull request by tim-dot-org**
Adds a simple view page to view reference guide content.

## Links

- spec: [updated spec](https://docs.google.com/document/d/1VZGle7QA-06ZNNxy2SkvA8Msrok8pyN7fuq5hzek888/edit)
- jira ticket: [PLAT-1353](https://codedotorg.atlassian.net/browse/PLAT-1353)

## Testing story

Added unit tests, tested manually on local server

## Follow-up work

Navigation menu on the left to be added.
Some kind of magic maybe to make the url include the entire category heirarchy instead of just the unique key part.

## Screenshots
Just rendering display name and content as markdown for now. Nav menu and breadcrumb links to be added.
image

### CODE REVIEW

File: dashboard/app/helpers/curriculum_helper.rb
Comment (tim-dot-org) [inline dashboard/app/helpers/curriculum_helper.rb:24]: Ran into a problem of not being able to just look up course version directly when there are extra "-"s in the key, so I borrowed this bit from vocabularies. Does this make sense to use for ref guides as well?

File: dashboard/app/models/reference_guide.rb
Comment (tim-dot-org) [inline dashboard/app/models/reference_guide.rb:28]: Currently using this, but maybe it makes more sense to just grab the unitgroup name. Is there an easy way to find that?

**Comment by dmcavoy** [comment]
Hey Tim - I will give this a look tomorrow. Could you add a screenshot of the change?

File: dashboard/app/helpers/curriculum_helper.rb
Comment (bethanyaconnor) [reply dashboard/app/helpers/curriculum_helper.rb:24]: Yes, I think so. The problem with CourseVersion is it doesn't have a good unique key to use in URLs, which is how this logic came about.

File: apps/Gruntfile.js
Comment (dmcavoy) [inline apps/Gruntfile.js]: Why is programming_expressions/show getting added here? Is it just a weird diff?

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: We probably don't want to require levelbuilder mode for show but we will for edit and update. Maybe add `, except [:show]` to the end of this line

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: can we try to use `load_and_authorize_resource` here? Also on that we update ability.rb for Reference Guides?

File: apps/Gruntfile.js
Comment (tim-dot-org) [reply apps/Gruntfile.js]: oh oops, yeah, that was from a merge conflict that i fixed incorrectly

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: It seems like it might be easier to manage this code going forward if we start the controller with 
`render :not_found if params[:course_course_name] && params[:key]` that way everything does not have to be inside of an if statement.

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: If we don't find an course_version_id it would probably be good to have a helpful error message.

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: I think this is another spot where if we don't get a reference guide it would be help to give a error message

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/reference_guides_controller.rb]: I think if you do the above error checking along the way you don't need this and the following lines. Rails does some magic to automatically call render :show without you needing to write it

File: dashboard/app/models/reference_guide.rb
Comment (dmcavoy) [reply dashboard/app/models/reference_guide.rb:28]: I don't think we can use UnitGroup name here because it could be either a UnitGroup or a Script. I think sticking with this for now makes sense

File: dashboard/test/controllers/reference_guides_controller_test.rb
Comment (dmcavoy) [inline dashboard/test/controllers/reference_guides_controller_test.rb]: You can create course offering and course version by calling `CourseOffering.add_course_offering(unit_group)` and that will make sure it is set up correctly for the unit_group

File: dashboard/app/models/ability.rb
Comment (dmcavoy) [inline dashboard/app/models/ability.rb:356]: Thank you for adding this update to ability.rb. This will make it so only levelbuilders can edit ReferenceGuides which is great. I think we also want to lock down ReferenceGuides to only be able to be viewed by instructor and participant audiences. I think we will need something like:

```
    can [:read], ReferenceGuide do |guide|
      course_or_unit = guide.course_version.content_root
      can?(:read, course_or_unit)
    end
```

I would add this near the line for Lesson `    can [:read, :show_by_id, :student_lesson_plan], Lesson do |lesson|`

File: dashboard/test/controllers/reference_guides_controller_test.rb
Comment (dmcavoy) [inline dashboard/test/controllers/reference_guides_controller_test.rb:43]: I think it would be good to add some tests to check who can access the show similar to the tests here: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/controllers/lessons_controller_test.rb#L89-L100

**Comment by dmcavoy** [comment]
Thanks for all the updates Tim! Things are looking good. I just had a couple more things I noticed when looking through things again but it is getting really close.

File: apps/Gruntfile.js
Comment (bethanyaconnor) [inline apps/Gruntfile.js]: nit: we learned recently that these teacher/student-facing pages should be in `codeStudioEntries` above, I think for performance reasons

File: dashboard/app/helpers/curriculum_helper.rb
Comment (bethanyaconnor) [reply dashboard/app/helpers/curriculum_helper.rb:24]: Thinking more about this -- I think this is fine for now but when we launch reference guides, we'll probably want to make sure this function isn't doing too many queries. This was previously used for a levelbuilder-facing page, so caching wasn't something we thought about. Definitely doesn't need to block this PR though

**Review by bethanyaconnor — APPROVED** [review]
LGTM after Dani's comments are all resolved

File: dashboard/app/models/ability.rb
Comment (tim-dot-org) [reply dashboard/app/models/ability.rb:356]: Will do! For my own understanding, what this is doing is checking the content root (UnitGroup or Script) and using the same permissions set for that. Is that right?

Comment (dmcavoy) [reply dashboard/app/models/ability.rb:356]: Yup - Its checking if the user can view the UnitGroup or Script that associated with the course version. More about the library can be found here: https://github.com/CanCanCommunity/cancancan/blob/develop/docs/README.md

Comment (dmcavoy) [reply dashboard/app/models/ability.rb:356]: Can you also add a test for all the ability settings in ability_test.rb?

File: dashboard/test/controllers/reference_guides_controller_test.rb
Comment (dmcavoy) [reply dashboard/test/controllers/reference_guides_controller_test.rb:43]: Thank you for adding these tests! Looks good!

**Review by dmcavoy — APPROVED** [review]
Thanks for all the follow up Tim. I left one more comment about adding some tests in ability_test.rb and then I think this PR is ready to go

File: dashboard/app/controllers/reference_guides_controller.rb
Comment (tim-dot-org) [inline dashboard/app/controllers/reference_guides_controller.rb:5]: settled back on just using `authorize_resource` since we're loading the resource manually with `find_reference_guide`
`id_param` added to get authorize to work correctly (see courses_controller)

File: dashboard/test/models/ability_test.rb
Comment (tim-dot-org) [inline dashboard/test/models/ability_test.rb]: unsure if this one should be `cannot?` it seems like an admin would still be able to view reference guides (esp if logged out can as well)

Comment (dmcavoy) [reply dashboard/test/models/ability_test.rb]: Good call out. Looks like you might need to also update here: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/ability.rb#L408

File: dashboard/test/models/ability_test.rb
Comment (dmcavoy) [inline dashboard/test/models/ability_test.rb]: It would be good to add a reference guide to one of the units with different audiences such as @public_facilitator_to_teacher_unit and test the ability to view that.

**Comment by dmcavoy** [comment]
Thanks again Tim for taking on the long tail of fixes here. Everything looks great!

**Comment by tim-dot-org** [comment]
> Thanks again Tim for taking on the long tail of fixes here. Everything looks great!

Thanks for reviewing it all! I learned a lot about cancan in the process

**Pull request merged**
